### DISCUSSION TITLE
Angle between cell surface normal and a line

### DISCUSSION THREAD

**Discussion by AsimAl-Sofi** [pyvista/pyvista#4286] [Q&A] [answered]
I would like to determine the angle between a line that goes through an observation point P and the cell face center point S AND the surface normal
Here is an example of a structured grid:

```python
out = pv.CylinderStructured(radius=(0,3),
theta_resolution=10,height=1,z_resolution=2)
cell_face=out.get_cell(5).get_face(3).cast_to_unstructured_grid()
point_P=out.get_cell(5).get_face(3).cast_to_unstructured_grid().cell_centers()
point_S = (5, 0, -1)
```
any suggestions to get this working?
<img width="424" alt="surface normal" src="https://user-images.githubusercontent.com/130053977/232333609-ab19359e-3245-41fc-8e28-e10c374145ec.PNG">

**Comment by tkoyama010** [comment]
Is adding #4266 features what you want to do?

**Reply by AsimAl-Sofi** [reply]
Hi @tkoyama010
There are some similarities but they are not the same

**Reply by AsimAl-Sofi** [reply]
Is there a way to determine the normal vector of the faces?

**Reply by MatthewFlamm** [reply]
https://docs.pyvista.org/version/stable/examples/01-filter/compute-normals.html

**Reply by AsimAl-Sofi** [reply]
@MatthewFlamm thanks for the link
I have read it before but I dont know how to use it in my case
For example, I want to get the normal vector of the following face surface:
out.get_cell(0).get_face(0)

**Reply by MatthewFlamm** [reply]
Ah you are right, this is not so clear, we only currently can cast to `UnstructuredGrid` for which you do not have a `compute_normals` filter.  This filter only works for `PolyData`.  Does this work?

```py
out.get_cell(0).get_face(0).cast_to_unstructured_grid().extract_surface().compute_normals()["Normals"][0]
```

  The first face (0 index) of most cells I checked seems to be either a point or a line and has no area, and therefore no normal.  When I check the second (1 index), it has a valid normal.

**Reply by MatthewFlamm** [reply]
I made #4291 so that in the future, we might be able to do something like, which is slightly more understandable.  This is only psuedo-code.

```py
out.get_cell(0).get_face(0).cast_to_polydata().compute_normals()["Normals"][0]
```

**Reply by AsimAl-Sofi** [reply]
@MatthewFlamm  thanks
Yes indeed I wanted something like the psuedo-code you mentioned but did not work since cast_to_polydata() does not work for cell.

**Reply by AsimAl-Sofi** [reply]
@MatthewFlamm I have checked your code provided above, and it is working
thank you for your help.
Is there an easy way to determine the angle between a line that goes through ab observation point and the surface center point and the normal vector?

**Reply by MatthewFlamm** [reply]
https://stackoverflow.com/questions/2827393/angles-between-two-n-dimensional-vectors-in-python/13849249#13849249

**Reply by AsimAl-Sofi** [reply]
I have also find the following link:
https://stackoverflow.com/questions/28260962/calculating-angles-between-line-segments-python-with-math-atan2
The method mentioned here calculates the angle between two lines which is exactly what I want.
However, for the normal vector I have only one point, i.e., the starting point, and a direction. Is there a way to determine the 'ending' point of the normal vector that has a magnitude of say 1 unit?

**Reply by MatthewFlamm** [reply]
I'd recommend the answer I provided, but they should be nearly equivalent.  The normal should already have a magnitude of 1 from vtk/pyvista unless it is degenerate.  If not, you can do something like `unit_normal = normal/np.linalg.norm(normal)`, assuming normal is a `np.ndarray`.

**Reply by AsimAl-Sofi** [reply]
@MatthewFlamm  thanks for your answer
In the link you have sent me, they assume that both vectors have a starting point at origin. In my case I have two vectors that do not start at origin. 
Using compute_normal method, it returns the direction of the face normal, but does not give the ending point of the normal vector.
here is my code:
import numpy as np
out = pv.CylinderStructured(radius=(0,3),
theta_resolution=10,height=1,
z_resolution=2)
box=out.get_cell(5).get_face(3).cast_to_unstructured_grid().extract_surface()
box.compute_normals(inplace=True)
centers = out.get_cell(5).get_face(3).cast_to_unstructured_grid().cell_centers().points
arrows = box['Normals']
pl = pv.Plotter()
pl.add_mesh(out.get_cell(5).get_face(3).cast_to_unstructured_grid(),color="g", opacity=0.6, show_edges=True, line_width=1)
pl.add_mesh(out.get_cell(5).get_face(3).cast_to_unstructured_grid().cell_centers(),color="r", point_size=18.0, render_points_as_spheres=True)
pl.add_mesh(out, opacity=0.6, show_edges=True, line_width=1)
pl.add_arrows(centers, arrows, color='black')
pl.add_points(arrows,color="g", point_size=18.0, render_points_as_spheres=True)
<img width="401" alt="surface normal4" src="https://user-images.githubusercontent.com/130053977/232801531-7b3fcda8-ee6c-4fcb-b331-17f7ec78e7d8.PNG">
pl.show()
As you can see, the green point does not corresponds to the normal vector ending point. 
Any suggestions how to determine this point?

**Reply by MatthewFlamm** [reply]
If you are talking about the vector from red to green point, it would be `pg-pr` where both `pg` and `pr` are `ndarray` with shape `(3,)`.  `pg` is the position of green point and similar for `pr` and red point.

**Reply by AsimAl-Sofi** [reply]
Hi
Iam talking about how to determine the black arrow's end point

**Reply by MatthewFlamm** [reply, marked as answer]
The end point is `pr + normal`.

**Reply by AsimAl-Sofi** [reply]
@MatthewFlamm thanks for your help
